### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by odagonzpfm1**
odagonzpfm1

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -4,130 +4,160 @@
...
 {
-    <#type name#> compare(<#type name#> a, <#type name#> b) //5
+    T* compare(T* a, T* b) 
     {
-        if( a->value < b->value ) return a;
-        if( a->value > b->value ) return b;
+        if (a == nullptr || b == nullptr) return nullptr;
+        if (a->distance < b->distance) return a;
+        if (a->distance > b->distance) return b;
         return nullptr;
```
Comment (matkatmusic) [inline main.cpp]: D.R.Y.
You have written `return nullptr;` twice in this function.
Revise the logic such that you only write it once.
- [x] revise

File: main.cpp [hunk 2]
```diff
@@ -4,130 +4,160 @@
...
-    std::cout << "the smaller one is << " << smaller->name << std::endl; //9
-    
-    U <#name3#>;
+    T firstSign(11, "A");
+    T secondSign(17, "B");
+
+    A f;
+    auto* smaller = f.compare(&firstSign, &secondSign);
+
+    std::cout << "the smaller one is << " << smaller->name << ".\nIf compare() returns a nullptr, then nothing will print out."<< std::endl;
```
Comment (matkatmusic) [inline main.cpp]: What do we always do before using a pointer?
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp (hunk 2 above)
Comment (odagonzpfm1) [reply main.cpp]: Check its not null

File: main.cpp (hunk 1 above)
Comment (odagonzpfm1) [reply main.cpp]: DRY up the code, however I still get a compiler warning for not having all paths covered. I think I'm covering all the paths (<, >, ==) in the logic plus nullptr, however still looks weird not having a "catch all" return at the bottom. 
How would you go about it?

File: main.cpp [hunk 3]
```diff
@@ -4,130 +4,169 @@
...
+    auto* smaller = f.compare(&firstSign, &secondSign);
+
+    if (smaller == nullptr) 
+    {
+        std::cout << "Smaller returned a nullptr. This may be due to any of the following possibilities: the T structs that are passed on are not properly initialized; or they point to a nullptr themselves; or the return value of the compare function is returning a nullptr." << std::endl;
+    }
+    else
+    {
+        std::cout << "the smaller one is << " << smaller->name << ".\nIf compare() returns a nullptr, then nothing will print out."<< std::endl;
+    }
```
Comment (matkatmusic) [inline main.cpp]:
```
if( smaller == nullptr )
{
    std::cout << "one or more arguments to the compare() function is a nullptr or all arguments' 'distance' member variable all have the same value." << std::endl;
}
else
{
    std::cout << "the smaller one is " << smaller->name << std::endl;
}
```
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**